### PR TITLE
Add drum pad instrument

### LINKED ISSUE #1460: Drum playing plugin
It has been mentioned in issue #1087 about adding a sample based drum machine, and issue #1396 about adding a drum synth. 

Tres has also suggested on the mailing list.

> I hate to describe other DAWs, but the drum sample in Ableton is very simple and effective and IIRC, it -uses it's default sampler instrument for each individual sample but IIRC, it also can allow other instrument types.

This is not so much an instrument, more of a multi instrument host. The idea is that each drum sound is played by it's own instrument. our new drum instrument, simply routes the note data to the assigned instrument.  ie the kick drum could use Kicker, the hats from Audio File Player, the toms form triple- oscillator. The choice of plugins would be down to the user.  This does work very well in practice.

How ever i am unsure of the implications of this, And would like to know any thoughts on the idea. or even if it is possible?

### PR DESCRIPTION
**Pull request by tresf**
@gi0e5b06 has written a drum-playing plugin.  The UI is rough, but it's a start.  Please contribute to this PR by pushing directly to `tresf/pads`.  If you don't have commit access, ask.

**Todo:**
* [ ] Reformat code
   * Use modern format
   * Revert unnecessary changes to `SampleBuffer`
   * Fix inconsistencies in formatting
* [ ] Decide on permanent name
* [ ] Create artwork
* [ ] Make UI intuitive, proposal here: #4542, thanks @RoxasKH!
* [ ] Ship at least one preset that works with common hardware/built-in samples
💡 *Please keep this listing up-to-date*

image

💡 *Please keep this screenshot up-to-date*

**Mockup proposal by @RoxasKH**
image

Closes #1460.

### CODE REVIEW

**Comment by qnebra** [comment]
Maybe name this PADS?

Intuitive UI, hmm.

**Comment by Spekular** [comment]
I get that "pad" in this case refers to a drum pad, but I don't think it's
a good name.

I don't see anything that makes this specifically designed for a drumpad as
opposed to being a generic drum sampler. This instrument is useful to
anyone, with or without a drumpad (even w/o any hardware).

As for UI, using the piano to select notes feels weird. What about a
scrolling vertical list of notes on the right and a list of samples on the
left? I can make a mockup later, but this would let you view several notes
and their assigned samples (maybe with some settings) at once, and feels
more intuitive to me.

**Comment by qnebra** [comment]
I think about image of drum pads, but it requiered ability to detect correct layers of pads in midi device. So @Spekular idea looks better, because my is too complicated.

**Comment by lukeoftheaura** [comment]
Maybe GUI wise something like a simplified drumkv1?

drumkv1

**Comment by tresf** [comment]
Please avoid things like "copy this design" or "I don't like the name".   The scrolling vertical list of implemented notes is going to be a hard requirement and will use existing LMMS controls inside a scrollpane.  The note selection is fine because it associates the pad feedback with the exact note that requires a sample. It will be in-addition-to the scrolling notes.

**Comment by Spekular** [comment]
I don't like the name because I think it's misleading, I don't see how that's not a valid critique. It'd be dumb to name TripleOscillator "SuperSawSynth", and this is also an overly-specific name.

As for the scrolling notes list, I don't see that anywhere in the report, so maybe I've missed it. I still don't think the piano should be used to select a note, as no other instruments behave that way. How should users play the instrument if the piano is used to select notes? If the piano is used both to play the instrument and select a note, will the UI update over and over when you play a pattern? I don't think this would necessarily be a performance concern, but it would be distracting/ugly.

FWIW, here's my mockup. Plugins are a bit narrower than I had in mind and it's a bit ugly, but I still tink it's a workable design:
image

At this size its already pretty cramped, but the ability to place multiple samples on one note (for randomization/velocity/etc.) is a potential enhancement.

**Comment by Spekular** [comment]
Just a mockup re:multiple samples on one note. Still ugly, not necessarily intuitive to anyone except me (heck, it might even confuse me if I actually had to use it, but it seems simple enough in theory).
image
Left/Right buttons switch between samples placed on one note (I suppose there'd have to be an empty slot all the time or a "+" icon to add one), Red/Blue select max/min velocity for that sample to play (would need to pop a dialogue box), Yellow is normalize (should be a toggle).

**Comment by qnebra** [comment]
Or maybe control this in pianoroll, as another mode of existing pianoroll, like Bitwig have.

**Comment by DeRobyJ** [comment]
@qnebra Having a different mode has some disadvantages.
Users will probably see their pads as midi notes, so using the easy to read pianoroll is a good thing imo.
We should focus on the UI for now, then we can think of creating a new kind of pianoroll for drums.
_________
**Strange idea (that deviates from this pull request)**
Now, I just want to ask if the following is a viable thing or not:
A fixed number of AudioFileProcessor Instrument editable instances _inside_ this plugin.

These instances would be editable, so we wouldn't need many controls.
We could have 4 tabs with 16 instances each, each one of them in little 50x50 pixel squares. 64 is the number of buttons on the two most famous pad midi devices.

padmockup
Basic mockup, squares can have the note they trigger written inside.

These squares have room for one standard knob and a led.

Basically I'm asking if it is possible to open an instance of a plugin inside this plugin, and using it not as a track but as an audio source only.

The user could put samples (loaded into AudioFileProcessor instances) or presets inside the squares, and then they could be edited by opening the instance of the plugin, similar to editing an instrument track.


I understand this is something completely different from this pull request. So back to it.

_________
**Mockup for this pull request**
Something like this:
padmockup2

My point is that I don't think it's that important to have a list of the samples, or to have those controls for the samples.
A fixed number of samples is ok imo (and in Ableton opinion).
Downside: this mockup only takes 32 samples into account.
**The user could, however, connect two instances of this plugin to their midi drumpad and change the base note.**
So they can have any multiple of 32 samples.
And yeah, with this idea we could ditch the tabs completely and have 8 samples per instance, but I'm not really a fan of this. It might work for live sessions, not for editing a drumloop with more than 8 samples.

**Controls**:
Volume and Panning can be edited inside the PianoRoll (or via Midi) by using Velocity and Panning control change.
Pitch isn't that important for drums, but we can actually have this one knob inside the squares. Let's say it's for tuning.

This kind of UI suggest the user that this plugin is intended from drum pads.
It's also easier to figure that you can drag and drop samples to the plugin, imo.

_______
Some name ideas:
Pad.io / Kit.io
Have you heard of these new browser games that use a .io domain, such as agar.io or slither.io?
These games have very simple gameplay mechanics, the main feature is the fact that many many players compete at the same time.
Similarly, our instrument has simple samples, the whole point is having them together in the same track.

_Kit_ty

**Comment by tresf** [comment]
I love the ideas but they're in the wrong thread  This is a pull request with very specific steps that need to be taken before it is merged.

Although we encourage progressive ideas, they're simply out of scope for this pull request unless somebody comes forward with code to back up the ideas.

**Comment by tresf** [comment]
Sorry, I missed this part... 

> Pad.io / Kit.io
> Have you heard of these new browser games that use a .io domain, such as agar.io or slither.io?
These games have very simple gameplay mechanics, the main feature is the fact that many many players compete at the same time.

Yes, very familiar with agario, slither, paper, etc.  They're games, this is a plugin.  We'll consider "Padio" and "Kitio", but the dot really has no place in a plugin name.  I feel comfortable saying this as the purchaser of two < 4 letter .io domains. :)

**Comment by Umcaruje** [comment]
I think we should move to simple, recognisable names, I think `Multisampler` or `Drum Machine` would be a nice name for this plugin. Though I would lean to Multisampler, since Drum Machine could be used for that more advanced plugin that allows loading other plugins that @DeRobyJ was talking about

**Comment by tresf** [comment]
Name idea from @sirbothersome "Conchpad".

**Comment by Spekular** [comment]
Somewhere in the message with kit.io and pad.io I thought I saw "kitty" as
well.

**Comment by tresf** [comment]
> Somewhere in the message with kit.io and pad.io I thought I saw "kitty" as well.

At the end, @DeRobyJ simply wrote `_Kit_ty`, but without context.  I'm going to use `Conchpad` for the short term I think.   It makes the logo work fairly well defined and has a nice ring to it.  If a better name is proposed before merging we can entertain that.  `Multisampler` and `Drum Machine` are just fine proposals too of course.  They're good because they match the descriptive named plugins like `TripleOscillator`, `AudioFileProcessor`, `SF2`, but we also have some creatively named ones like `Xpressive`, `Kicker`, `Watsyn`, etc so I'm choosing the later.

File: plugins/PadsGDX/PadsGDX.cpp [hunk 1]
```diff
@@ -418,7 +418,7 @@ bool PadsGDX::checkPointBounds(int key) {
 		return false;
 	}
 
-	SampleBuffer* sample = m_sampleBuffer[key];
+	SampleBuffer* sample = m_sampleBuffer[key]; Q_UNUSED(sample);
```
Comment (lukas-w) [inline plugins/PadsGDX/PadsGDX.cpp]: Why not just remove this line?

**Comment by NETMANSKY** [comment]
In my modest opinion the creation of DrumPAD instrument is a brilliant idea! But I am more that sure it must support multichannel routing!!! Each sample must be proccessed in assigned mixer channel. Because proccessing of kick, snare or hats/perc is completely different.

**Comment by tresf** [comment]
> Each sample must be proccessed in assigned mixer channel. Because proccessing of kick, snare or hats/perc is completely different.

This is a valid point.  We'll consider this for a future release, thanks.

**Comment by ahstro** [comment]
Is this still being worked on? If not/while we wait, what is the recommended way to get this functionality?

**Comment by RoxasKH** [comment]
I just want to notice that my work was made for a general multisampler and not for something like this. The idea is like the same, but since this was supposed to be a drumpad, maybe at the beginning you could use this
pads
Also because my mockup don't fit the normal lmms instrument GUI dimension, and that was never done, so it should be difficult to program.
I let the chioce to you anyway

**Comment by RoxasKH** [comment]
Well, i had time so i think i made the final mockup, at least for a drumpad (i already made the multisampler one).
I also think i covered all the functionality a drumpad need, and it was kind of difficult to make this in the little default instrument size of lmms instruments.
So here it is (DeRobyJ inspired this a little, it still need some tweak)
pads2

Let's start explaining how i figured out it would work.
So the top are the 8 pads theirselves and the logo. All pads have a little label with a number and are linked to a specific key 
1-C4, 2-C#4, 3-D4, 4-D#4, 5-E4, 6-F4, 7-F#4, 8-G
In this way you can play a pad by clicking on it or just playing the key it's linked to. I choose these keys, so that editing your midi pattern in piano-roll should be easier.
You can also rename a pad as you want by right clicking on the text box over them (default names are the notes they're connected to). As you can see when a pad is played is lighted up.

The down part of the plugins focus on sample editing, to be more precise on single-sample editing. That means you can edit every single sample in every pad in a different way. You can surf through the pads using the menu on the right where are listed all the pads with their number (Pad1, Pad2, Pad3... and so, the number is the little label one) or by ctrl-clicking on the pad sample you want to edit.
As you can see there's a visual of the sample (like the original and the AFP one), a reverse button, ADSR, volume and panning knobs, and the sample start and end ones. You can also mute a pad by clicking on the power button (i dunno why but i left the mute/solo thing i put over the visual, it's also not so good to see, that shouldn't be there, sorry), and as you can see it will get grey.
On the left it's written the original name of the sample.

But i haven't already said the most important thing and so how to load your sample in a pad. There are two ways, that are drag&dropping it over the pad, or choosing the pad by the scrolldown menu and clicking on the green folder on the left.

As a drumpad, so studied to play some percussive sounds, i decided to remove some things, like the loop button and the stutter one to save space for other more crucial things. I would have added a reverb knob but it's not so needed on single samples, and i'm also sorry for the fact you can't directly change the pitch of the sample, but it's still really good in my opinion.

Hope this will be programmed in the future, regard, RoxasKH (lol, my name is just so random, my artist name is H4CKTON3). I think this is all, maybe we should tag umcaruje to review the graphical work.

**Comment by NETMANSKY** [comment]
@RoxasKH 🥇 It's fine. Good stuff! But maybe pad linking to a static key is not right way?  Probable each pad linking to key must be freely same as root key on keyboard in AFP. Maybe is it better way? It will let to layering a samples for example. :)
In my opinion each pad must route to a selected Fx channel of mixer. And maybe need to increase a number of pads to 16?
Anyway you drawed good design 👍

**Comment by RoxasKH** [comment]
@NETMANSKY What are you talking about is just difficult af to code (we also need to think that), and the idea of an instrument that will fill 8 FX channels is a bit weird to me, also why coding that when you can obtain the thing you want to do, just using 8 AFP? Also how would you do the piano-roll? A different one for every pad? As i said if you want several piano rolls, use different AFPs. Also remember this is studied to play percussive sounds, for instruments like piano or violins i designed the multisampler you can see in the first post (LMMSampler).
I think it's too much and i never see a drumpad plug which use more than 1 fx channel.
As for the 16 pads, guy that's not possible in the default lmms instrument size (what could i remove? I had to use little knobs to make it fit right)

**Comment by NETMANSKY** [comment]
@RoxasKH I do not insist and agree with you. But traditionally VST pads have the ability to assign mixer channels in the settings of each pad. For example: the kick has its own processing chain in the mixer channel but snare and wooden percussion has another chain in another channel. Hats and cymbals also require their processing. And AFP had nothing to do with it. Agree me that the multisampler and pads are the same. The required keys, which are assigned in both instruments, are typed exactly the same in piano roll. Isn't it?
Typically VST pads have multichannel support for separate processing of sounds as a few separated AFP but  multisampler plays the sound in one channel as SF2 player.
Anyway you work hard and I srong believe that our team of developers and designers will succeed! Just do your job and life itself will tell the right decision!

**Comment by RoxasKH** [comment]
I agree with you with all, different processing of drums and the comfort of having all drums in only one instance. But i don't agree when you say that multisampler and pads are the same. Yeah you can use the Multisampler as a DrumPad, but not a drumpad as a multisampler. Have you ever used something like FL Directwave or OneSmallClue Grace? That's a multisampler, studied also to load sfz files.
Even if i agree with you, as i said it before (and you agreed wit me i guess) could be hard to code so at the moment i will appreciate it even only as something with which you can save drumsets (also i put some basic editing for separate pads as you can see).
Also i'm not a designer or somthing, it's only an hobby, never did a school or something similar. I only commit myself a lot, so it's nice to see that my work is appreciated!

**Comment by NETMANSKY** [comment]
And 

@RoxasKH  > Have you ever used something like FL Directwave or OneSmallClue Grace? That's a multisampler, studied also to load sfz files.
👍 
And also Sforzando, TX16W, TAL Sampler... Of course. I know the subject. But these are VST and LMMS working hard with it. 
If we talk about pads, I just want to help at least with my vision and do not pretend to be true. If the team has difficulties with the implementation of multichannel and midi routing or another features, then there is nothing wrong with that. In the next version it will succeed. :-) 

 P.S. The most important thing is that community members will hear each other

**Comment by tresf** [comment]
Closing as it's been stale for over a year.  Someone can resurrect this at their own will.

**Pull request closed without merging**